### DISCUSSION TITLE
Axis camera RTSP stream feeding into RF-DETR with OpenCV handling the display and capture.

### DISCUSSION THREAD

**Discussion by vivekatoffice** [AxisCommunications/discussions#1262] [General]
## Sample1.py ([RF-DETR](https://huggingface.co/papers/2407.17140) proposes a Receptive Field Detection Transformer (DETR))

Reference: ([Code Link](https://huggingface.co/docs/transformers/main/model_doc/rf_detr))

### Input Image: cardemo.png
<img width="1536" height="1024" alt="cardemo" src="https://github.com/user-attachments/assets/f34f9568-b6e8-4005-ba8a-ad41cc30793c" />

```py
import os
os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"

from transformers import AutoImageProcessor, AutoModelForObjectDetection
from PIL import Image
import supervision as sv
import requests
import torch

image_path = "cardemo.png"
image = Image.open(image_path).convert("RGB")

image_processor = AutoImageProcessor.from_pretrained("Roboflow/rf-detr-medium")
model = AutoModelForObjectDetection.from_pretrained("Roboflow/rf-detr-medium", device_map="auto")

inputs = image_processor(images=image, return_tensors="pt").to(model.device)

with torch.no_grad():
    outputs = model(**inputs)

results = image_processor.post_process_object_detection(
    outputs, target_sizes=torch.tensor([image.size[::-1]]), threshold=0.3
)[0]

detections = sv.Detections.from_transformers(
    transformers_results=results, id2label=model.config.id2label
)

box_annotator = sv.BoxAnnotator()
label_annotator = sv.LabelAnnotator()

annotated_image = image.copy()
annotated_image = box_annotator.annotate(annotated_image, detections)
annotated_image = label_annotator.annotate(annotated_image, detections)

sv.plot_image(annotated_image)

# Save the annotated image to verify results
if isinstance(annotated_image, Image.Image):
    annotated_image.save("output.jpg")
    print("Saved annotated image to output.jpg")
else:
    import numpy as np
    if isinstance(annotated_image, np.ndarray):
        # Convert BGR/RGB numpy array back to PIL Image and save
        # supervision usually annotates on BGR, but if input was PIL it might be RGB.
        # Let's check if we need to convert BGR to RGB.
        # If it was converted from PIL, it might be in RGB or BGR.
        # To be safe, we can just save it or convert BGR to RGB.
        # Usually supervision expects BGR numpy array.
        # Let's try converting it or saving directly.
        img_to_save = Image.fromarray(annotated_image)
        img_to_save.save("output.jpg")
        print("Saved annotated image numpy array to output.jpg")

```
## Output Image:

<img width="1536" height="1024" alt="output" src="https://github.com/user-attachments/assets/60d92f48-d4c5-4006-aa3d-dc919b7f78ca" />

## Sample2.py

```py
import os
# Configure HF mirror to avoid connection timeout/reset
os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"

import threading
import time
import argparse
from PIL import Image
import cv2
import torch
import supervision as sv
from transformers import AutoImageProcessor, AutoModelForObjectDetection

class RTSPStreamReader:
    """
    Threaded reader to grab frames from the RTSP stream continuously.
    Prevents OpenCV buffer accumulation and reduces stream latency lag.
    """
    def __init__(self, rtsp_url):
        self.rtsp_url = rtsp_url
        self.cap = cv2.VideoCapture(self.rtsp_url)
        self.frame = None
        self.ret = False
        self.stopped = False
        self.lock = threading.Lock()
        
    def start(self):
        t = threading.Thread(target=self.update, args=())
        t.daemon = True
        t.start()
        return self
        
    def update(self):
        while not self.stopped:
            if not self.cap.isOpened():
                time.sleep(0.5)
                # Re-try opening
                self.cap = cv2.VideoCapture(self.rtsp_url)
                continue
            
            ret, frame = self.cap.read()
            if not ret:
                time.sleep(0.01)
                continue
                
            with self.lock:
                self.frame = frame
                self.ret = ret
                
    def read(self):
        with self.lock:
            if self.frame is None:
                return False, None
            return self.ret, self.frame.copy()
            
    def stop(self):
        self.stopped = True
        if self.cap.isOpened():
            self.cap.release()


def main():
    parser = argparse.ArgumentParser(description="Run DETR object detection on Axis RTSP camera stream.")
    parser.add_argument("--ip", default="192.168.68.106", help="IP address of the Axis camera")
    parser.add_argument("--user", help="RTSP username")
    parser.add_argument("--password", help="RTSP password")
    parser.add_argument("--threshold", type=float, default=0.3, help="Detection confidence threshold")
    parser.add_argument("--save_interval", type=float, default=2.0, help="Interval in seconds to save frames for verification")
    parser.add_argument("--headless", action="store_true", help="Run in headless mode (no GUI window, saves stream_output.jpg instead)")
    args = parser.parse_args()

    # Build the RTSP URL
    if args.user and args.password:
        rtsp_url = f"rtsp://{args.user}:{args.password}@{args.ip}/axis-media/media.amp?resolution=640x480"
    else:
        rtsp_url = f"rtsp://{args.ip}/axis-media/media.amp?resolution=640x480"

    print("Loading DETR Object Detection Model...")
    image_processor = AutoImageProcessor.from_pretrained("Roboflow/rf-detr-medium")
    model = AutoModelForObjectDetection.from_pretrained("Roboflow/rf-detr-medium", device_map="auto")
    print(f"Model loaded onto device: {model.device}")

    box_annotator = sv.BoxAnnotator()
    label_annotator = sv.LabelAnnotator(text_position=sv.Position.CENTER)

    print(f"Connecting to RTSP Stream: {rtsp_url} ...")
    reader = RTSPStreamReader(rtsp_url).start()

    # Wait for the first frame
    print("Waiting for stream frames...")
    for _ in range(30):
        ret, frame = reader.read()
        if ret:
            break
        time.sleep(0.2)
    else:
        print("Warning: No initial frames received from the RTSP stream. Check if the camera is online and URL is correct.")

    last_save_time = 0
    frame_count = 0
    
    print("Starting detection loop. Press Ctrl+C in terminal or 'q' in window to exit.")
    try:
        while True:
            start_time = time.time()
            ret, frame = reader.read()
            if not ret or frame is None:
                time.sleep(0.01)
                continue

            # Convert frame to RGB for PyTorch/Hugging Face image processor
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            pil_image = Image.fromarray(frame_rgb)

            # Preprocess and inference
            inputs = image_processor(images=pil_image, return_tensors="pt").to(model.device)
            with torch.no_grad():
                outputs = model(**inputs)

            # Post-process detections
            results = image_processor.post_process_object_detection(
                outputs, target_sizes=torch.tensor([pil_image.size[::-1]]), threshold=args.threshold
            )[0]

            detections = sv.Detections.from_transformers(
                transformers_results=results, id2label=model.config.id2label
            )

            # Create labels with class name and confidence score
            labels = [
                f"{model.config.id2label[class_id]} {confidence:.2f}"
                for class_id, confidence in zip(detections.class_id, detections.confidence)
            ]

            # Annotate the frame (on a copy of the original BGR frame)
            annotated_frame = box_annotator.annotate(frame.copy(), detections)
            annotated_frame = label_annotator.annotate(annotated_frame, detections, labels=labels)

            # Compute FPS
            fps = 1.0 / (time.time() - start_time)
            
            # Draw FPS on frame
            cv2.putText(
                annotated_frame, 
                f"FPS: {fps:.2f}", 
                (20, 40), 
                cv2.FONT_HERSHEY_SIMPLEX, 
                1, 
                (0, 255, 0), 
                2
            )

            # Print detection summary
            detected_labels = [model.config.id2label[class_id] for class_id in detections.class_id]
            print(f"[Frame {frame_count}] Detected: {detected_labels} | Inference speed: {fps:.2f} FPS")

            # Try displaying using OpenCV GUI (if not in headless mode)
            if not args.headless:
                try:
                    cv2.imshow("Axis Camera - DETR Object Detection", annotated_frame)
                    if cv2.waitKey(1) & 0xFF == ord('q'):
                        break
                except cv2.error:
                    pass

            # Periodically save latest frame to disk
            current_time = time.time()
            if current_time - last_save_time >= args.save_interval:
                cv2.imwrite("stream_output.jpg", annotated_frame)
                print(f"--> Saved current frame to stream_output.jpg (Headless Verification)")
                last_save_time = current_time

            frame_count += 1

    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        reader.stop()
        cv2.destroyAllWindows()
        print("RTSP stream disconnected and windows closed.")

if __name__ == "__main__":
    main()


```

### Oputput: 

[LinkedIn Post](https://www.linkedin.com/posts/vivektarwan_computervision-ai-aiagents-ugcPost-7465662949597278208-3QLS/?utm_source=share&utm_medium=member_desktop&rcm=ACoAAAe-qoMBDNdYyEkT-oXK3Rfd-FhxUC7AOOs)

## Sample3.py

```py
from transformers import AutoImageProcessor, AutoModelForInstanceSegmentation
from PIL import Image
import supervision as sv
import requests
import torch

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

image_processor = AutoImageProcessor.from_pretrained("Roboflow/rf-detr-seg-medium")
model = AutoModelForInstanceSegmentation.from_pretrained("Roboflow/rf-detr-seg-medium", device_map="auto")

inputs = image_processor(images=image, return_tensors="pt").to(model.device)

with torch.no_grad():
    outputs = model(**inputs)

results = image_processor.post_process_instance_segmentation(
    outputs, target_sizes=[image.size[::-1]], threshold=0.3
)[0]

detections = sv.Detections.from_transformers(
    transformers_results=results, id2label=model.config.id2label
)

mask_annotator = sv.MaskAnnotator()
label_annotator = sv.LabelAnnotator()

annotated_image = image.copy()
annotated_image = mask_annotator.annotate(annotated_image, detections)
annotated_image = label_annotator.annotate(annotated_image, detections)

sv.plot_image(annotated_image)

annotated_image.save("outputannoted.jpg")
print("Saved annotated image numpy array to output.jpg")

```

## Output Image:

<img width="640" height="480" alt="outputannoted" src="https://github.com/user-attachments/assets/658b9f1f-f1bf-42d0-b045-94f288b487d5" />

## Sample3.py

```py
import os
# Configure HF mirror to avoid connection timeout/reset
os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"

import threading
import time
import argparse
import numpy as np
from PIL import Image
import cv2
import torch
import supervision as sv
from transformers import AutoImageProcessor, AutoModelForInstanceSegmentation


class RTSPStreamReader:
    """
    Threaded reader to grab frames from the RTSP stream continuously.
    Prevents OpenCV buffer accumulation and reduces stream latency lag.
    """
    def __init__(self, rtsp_url):
        self.rtsp_url = rtsp_url
        self.cap = cv2.VideoCapture(self.rtsp_url)
        self.frame = None
        self.ret = False
        self.stopped = False
        self.lock = threading.Lock()

    def start(self):
        t = threading.Thread(target=self.update, args=())
        t.daemon = True
        t.start()
        return self

    def update(self):
        while not self.stopped:
            if not self.cap.isOpened():
                time.sleep(0.5)
                self.cap = cv2.VideoCapture(self.rtsp_url)
                continue

            ret, frame = self.cap.read()
            if not ret:
                time.sleep(0.01)
                continue

            with self.lock:
                self.frame = frame
                self.ret = ret

    def read(self):
        with self.lock:
            if self.frame is None:
                return False, None
            return self.ret, self.frame.copy()

    def stop(self):
        self.stopped = True
        if self.cap.isOpened():
            self.cap.release()


def build_detections(results, id2label):
    """
    Manually construct sv.Detections from instance segmentation results.

    post_process_instance_segmentation returns a dict with:
      - 'segmentation': a 2D tensor where each pixel holds a segment id
      - 'segments_info': list of dicts with 'id', 'label_id', 'score'

    We convert that into masks, bounding boxes, scores, class_ids, and labels
    so supervision can annotate them cleanly.
    """
    seg_map = results["segmentation"]
    segments_info = results["segments_info"]

    if not segments_info:
        return sv.Detections.empty()

    masks = []
    scores = []
    class_ids = []
    labels = []

    for seg in segments_info:
        # Binary mask for this segment
        if torch.is_tensor(seg_map):
            mask = (seg_map == seg["id"]).cpu().numpy()
        else:
            mask = (seg_map == seg["id"])

        masks.append(mask)
        scores.append(seg["score"])
        class_ids.append(seg["label_id"])
        labels.append(id2label.get(seg["label_id"], "unknown"))

    masks_np = np.array(masks)

    return sv.Detections(
        xyxy=sv.mask_to_xyxy(masks_np),
        mask=masks_np,
        confidence=np.array(scores),
        class_id=np.array(class_ids),
        data={"class_name": labels},
    )


def main():
    parser = argparse.ArgumentParser(
        description="Run RF-DETR instance segmentation on an Axis RTSP camera stream."
    )
    parser.add_argument("--ip", default="192.168.68.106", help="IP address of the Axis camera")
    parser.add_argument("--user", help="RTSP username")
    parser.add_argument("--password", help="RTSP password")
    parser.add_argument("--threshold", type=float, default=0.3, help="Detection confidence threshold")
    parser.add_argument("--save_interval", type=float, default=2.0,
                        help="Interval in seconds to save frames for verification")
    parser.add_argument("--headless", action="store_true",
                        help="Run in headless mode (no GUI window, saves stream_output.jpg instead)")
    args = parser.parse_args()

    # Build the RTSP URL
    if args.user and args.password:
        rtsp_url = (
            f"rtsp://{args.user}:{args.password}@{args.ip}"
            f"/axis-media/media.amp?resolution=640x480"
        )
    else:
        rtsp_url = f"rtsp://{args.ip}/axis-media/media.amp?resolution=640x480"

    # ── Model loading ─────────────────────────────────────────────────
    print("Loading RF-DETR Instance Segmentation Model...")
    model_id = "Roboflow/rf-detr-seg-medium"
    image_processor = AutoImageProcessor.from_pretrained(model_id)
    model = AutoModelForInstanceSegmentation.from_pretrained(model_id, device_map="auto")
    print(f"Model loaded onto device: {model.device}")

    id2label = model.config.id2label

    # ── Annotators ────────────────────────────────────────────────────
    mask_annotator = sv.MaskAnnotator(opacity=0.4)
    box_annotator = sv.BoxAnnotator()
    label_annotator = sv.LabelAnnotator(text_position=sv.Position.CENTER)

    # ── RTSP stream ───────────────────────────────────────────────────
    print(f"Connecting to RTSP Stream: {rtsp_url} ...")
    reader = RTSPStreamReader(rtsp_url).start()

    print("Waiting for stream frames...")
    for _ in range(30):
        ret, frame = reader.read()
        if ret:
            break
        time.sleep(0.2)
    else:
        print(
            "Warning: No initial frames received from the RTSP stream. "
            "Check if the camera is online and URL is correct."
        )

    last_save_time = 0
    frame_count = 0

    print("Starting segmentation loop. Press Ctrl+C in terminal or 'q' in window to exit.")
    try:
        while True:
            start_time = time.time()
            ret, frame = reader.read()
            if not ret or frame is None:
                time.sleep(0.01)
                continue

            # Convert BGR → RGB → PIL for the HF image processor
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            pil_image = Image.fromarray(frame_rgb)

            # ── Inference ─────────────────────────────────────────────
            inputs = image_processor(images=pil_image, return_tensors="pt").to(model.device)
            with torch.no_grad():
                outputs = model(**inputs)

            # ── Post-process as instance segmentation ─────────────────
            results = image_processor.post_process_instance_segmentation(
                outputs,
                target_sizes=[pil_image.size[::-1]],
                threshold=args.threshold,
            )[0]

            # Build supervision Detections (masks + boxes + labels)
            detections = build_detections(results, id2label)

            # Readable labels with confidence
            labels = [
                f"{id2label.get(cid, 'unknown')} {conf:.2f}"
                for cid, conf in zip(detections.class_id, detections.confidence)
            ] if len(detections) > 0 else []

            # ── Annotate ──────────────────────────────────────────────
            annotated_frame = frame.copy()
            annotated_frame = mask_annotator.annotate(annotated_frame, detections)
            annotated_frame = box_annotator.annotate(annotated_frame, detections)
            annotated_frame = label_annotator.annotate(annotated_frame, detections, labels=labels)

            # FPS overlay
            fps = 1.0 / (time.time() - start_time)
            cv2.putText(
                annotated_frame,
                f"FPS: {fps:.2f}",
                (20, 40),
                cv2.FONT_HERSHEY_SIMPLEX,
                1,
                (0, 255, 0),
                2,
            )

            # Terminal log
            detected_labels = [id2label.get(cid, "?") for cid in detections.class_id]
            print(f"[Frame {frame_count}] Detected: {detected_labels} | {fps:.2f} FPS")

            # Display (unless headless)
            if not args.headless:
                try:
                    cv2.imshow("Axis Camera - RF-DETR Instance Segmentation", annotated_frame)
                    if cv2.waitKey(1) & 0xFF == ord("q"):
                        break
                except cv2.error:
                    pass

            # Periodic save
            current_time = time.time()
            if current_time - last_save_time >= args.save_interval:
                cv2.imwrite("stream_output.jpg", annotated_frame)
                print("--> Saved current frame to stream_output.jpg")
                last_save_time = current_time

            frame_count += 1

    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        reader.stop()
        cv2.destroyAllWindows()
        print("RTSP stream disconnected and windows closed.")


if __name__ == "__main__":
    main()
```
## Output Image: 
```txt
PS C:\Vivek\AxisTIP\API_MDs\rf_detr> python .\Sample4.py --ip 195.60.68.14:33111 --user VLTuser --password xXxXxXxX
```

<img width="640" height="480" alt="stream_output" src="https://github.com/user-attachments/assets/60b56053-2d08-4873-9306-9a9ed445d317" />

**Comment by github-actions** [comment]
*This automatically generated reply acts as a friendly reminder.*

*Answers to your questions will most often come from the community, from developers like yourself. You will, from time to time, find that Axis employees answers some of the questions, but this is not a guarantee. Think of the discussion forum as a complement to other support channels, not a replacement to any of them.*

*If your question remains unanswered for a period of time, please revisit it to see whether it can be improved by following the guidelines listed in [Axis support guidelines](https://github.com/orgs/AxisCommunications/discussions/13).*
